File: documentation/manual/working/scalaGuide/advanced/http/code/EssentialFilter.scala
package scalaguide.advanced.filters.essential

// #essential-filter-example
import play.api.Logger
import play.api.mvc._
import play.api.libs.concurrent.Execution.Implicits.defaultContext

class LoggingFilter extends EssentialFilter {
  def apply(nextFilter: EssentialAction) = new EssentialAction {
    def apply(requestHeader: RequestHeader) = {

      val startTime = System.currentTimeMillis

      nextFilter(requestHeader).map { result =>

        val endTime = System.currentTimeMillis
        val requestTime = endTime - startTime

        Logger.info(s"${requestHeader.method} ${requestHeader.uri}" +
          s" took ${requestTime}ms and returned ${result.header.status}")
        result.withHeaders("Request-Time" -> requestTime.toString)

      }
    }
  }
}
// #essential-filter-example

File: framework/src/play/src/main/java/play/mvc/EssentialFilter.java
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <https://www.lightbend.com>
 */
package play.mvc;

public abstract class EssentialFilter implements play.api.mvc.EssentialFilter {
    public abstract EssentialAction apply(play.mvc.EssentialAction next);

    @Override
    public play.mvc.EssentialAction apply(play.api.mvc.EssentialAction next) {
        return apply(next.asJava());
    }

    @Override
    public EssentialFilter asJava() {
        return this;
    }
}
